File: sast-js/case/completeness/single_app_tracing/class/complex_object/cross_class_001_T.js
// evaluation information start
// real case = true
// evaluation item = 完整度->单应用跟踪完整度->接口与类->复杂对象
// scene introduction = 跨类访问-成员变量
// level = 2
// bind_url = completeness/single_app_tracing/class/complex_object/cross_class_001_T
// evaluation information end
const { execSync } = require('child_process');


function cross_class_001_T(__taint_src) {
  class A {
    constructor() {
      this.data = __taint_src;
    }

    getData() {
      return this.data;
    }
  }

  class B {
    constructor() {
      this.data = new A().getData();
    }
  }

  let obj = new B();
  __taint_sink(obj);
}

function __taint_sink(o) {
  execSync(o);
}

const taint_src = "taint_src_value";

cross_class_001_T(taint_src);

File: sast-python2/case/completeness/single_app_tracing/class/complex_object/cross_class_001_T.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = true
# evaluation item = 完整度->单应用跟踪完整度->接口与类->复杂对象
# scene introduction = 跨类访问-成员变量
# level = 2
# bind_url = completeness/single_app_tracing/class/complex_object/cross_class_001_T
# evaluation information end
import os


def cross_class_001_T(taint_src):
    class A(object):
        def __init__(self, taint_src):
            self.data = taint_src

        def get_data(self):
            return self.data

    class B(object):
        def __init__(self, taint_src):
            a_instance = A(taint_src)
            self.data = a_instance.get_data()

    obj = B(taint_src)
    taint_sink(obj.data)


def taint_sink(o):
    os.system(o)


if __name__ == u'__main__':
    taint_src = u"taint_src_value"
    cross_class_001_T(taint_src)

